homeassistant/components/bluesound/media_player.py

```diff
@@ -829,83 +812,118 @@ def is_master(self):
 
     @property
     def shuffle(self):
         """Return true if shuffle is active."""
         return self._status.get("shuffle", "0") == "1"
 
-    async def async_join(self, master):
-        """Join the player to a group."""
-        master_device = [
-            device
-            for device in self.hass.data[DATA_BLUESOUND]
-            if device.entity_id == master
-        ]
-
-        if master_device:
-            _LOGGER.debug(
-                "Trying to join player: %s to master: %s",
-                self.id,
-                master_device[0].id,
-            )
+    @property
+    def group_members(self) -> list[str] | None:
+        """Return a list of entity_ids, which belong to the group of self."""
+        if self._group_list is None or not self.is_master:
+            return None
 
-            await master_device[0].async_add_slave(self)
-        else:
-            _LOGGER.error("Master not found %s", master_device)
+        return [entity.entity_id for entity in self._group_list]
 
     @property
     def extra_state_attributes(self):
         """List members in group."""
         attributes = {}
-        if self._group_list:
-            attributes = {ATTR_BLUESOUND_GROUP: self._group_list}
-
         attributes[ATTR_MASTER] = self._is_master
 
         return attributes
 
     def rebuild_bluesound_group(self):
         """Rebuild the list of entities in speaker group."""
-        if self._group_name is None:
+
+        if not self.is_master:
+            return None
+
+        # players not initialized this method will be triggered later
+        if self.hass is None:
             return None
 
-        bluesound_group = []
+        slaves = []
+        slave_node = self._sync_status.get("slave")
+        nodes = slave_node if isinstance(slave_node, list) else [slave_node]
 
-        device_group = self._group_name.split("+")
+        for slave in nodes:
+            slave_id = slave.get("@id")
+            slave_port = slave.get("@port")
+            slaves.append(f"{slave_id}:{slave_port}")
 
-        sorted_entities = sorted(
-            self._hass.data[DATA_BLUESOUND],
-            key=lambda entity: entity.is_master,
-            reverse=True,
+        bluesound_group: list[BluesoundPlayer] = []
+        bluesound_group.append(self)
+
+        bluesound_group.extend(
+            device for device in self.hass.data[DATA_BLUESOUND] if device.id in slaves
         )
-        bluesound_group = [
-            entity.name
-            for entity in sorted_entities
-            if entity.bluesound_device_name in device_group
-        ]
 
         return bluesound_group
 
-    async def async_unjoin(self):
-        """Unjoin the player from a group."""
-        if self._master is None:
-            return
+    async def async_join_players(self, group_members: list[str]) -> None:
+        """Join `group_members` as a player group with the current player."""
+
+        devices = [
```

Comment:
We shouldn't store the entities in a shared container. If we need to communicate with another entity we should use our dispatch helper.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/dispatcher.py

It's ok to store the entity ids of the group members.